internal/service/event_listener.go

```diff
@@ -0,0 +1,116 @@
+package service
+
+import (
+	"fmt"
+
+	"git.sr.ht/~spc/go-log"
+	"github.com/coreos/go-systemd/v22/dbus"
+)
+
+const (
+	// servicePostfixLength is the length of the string ".service". This value is used to extract the workload name from the service
+	// coming from systemd
+	servicePostfixLength = 8
+
+	EventStarted EventType = "started"
+	EventStopped EventType = "stopped"
+
+	// To avoid confusion, we match the action verb tense coming from systemd sub state, even though it is not consistent
+	// The only addition to the list is the "removed" case, which we use when a service is removed
+	// from the system and we receive an empty unitStatus object.
+
+	//The service is running
+	running unitSubState = "running"
+	//The service is starting, probably due to a restart.
+	//In this case we have to notify the observers that the service is not yet up and is loading. A subsequent 'running' event will proceed.
+	start unitSubState = "start"
+	//The service has been stopped, due to the container being killed. Systemd will restart the service.
+	stop unitSubState = "stop"
+	//The service is dead, due to the container being stopped for instance.
+	dead unitSubState = "dead"
+	//The service failed to start
+	failed unitSubState = "failed"
+	//The service has been removed, no unit state is returned from systemd
+	removed unitSubState = "removed"
+)
+
+type unitSubState string
+
+type EventType string
+
+type Event struct {
+	WorkloadName string
+	Type         EventType
+}
+
+type EventListener struct {
+	observerCh chan *Event
+	set        *dbus.SubscriptionSet
+	dbusCh     <-chan map[string]*dbus.UnitStatus
+	dbusErrCh  <-chan error
+}
+
+func NewEventListener(observerCh chan *Event) *EventListener {
```

Comment:
being the producer, I would own this channel. I can make sure it is properly initialized

Reply:
Agreed. Let me refactor this.

Reply:
Done. Please take a look.